--- a/drivers/gpu/drm/panel/panel-himax-hx8279.c
+++ b/drivers/gpu/drm/panel/panel-himax-hx8279.c
@@ -122,6 +122,7 @@
 	bool skip_goa_timing;
 	bool skip_goa_even_timing;
 	bool skip_mipi_timing;
+	enum drm_panel_orientation orientation;
 };
 
 struct hx8279_panel_mode {
@@ -755,12 +756,20 @@
 	return hx->desc->num_modes;
 }
 
+static enum drm_panel_orientation hx8279_get_orientation(struct drm_panel *panel)
+{
+	struct hx8279 *ctx = to_hx8279(panel);
+
+	return ctx->orientation;
+}
+
 static const struct drm_panel_funcs hx8279_panel_funcs = {
 	.disable = hx8279_disable,
 	.unprepare = hx8279_unprepare,
 	.prepare = hx8279_prepare,
 	.enable = hx8279_enable,
 	.get_modes = hx8279_get_modes,
+	.get_orientation = hx8279_get_orientation,
 };
 
 static int hx8279_init_vregs(struct hx8279 *hx, struct device *dev)
@@ -1037,6 +1046,12 @@
 		return dev_err_probe(dev, PTR_ERR(hx->reset_gpio),
 				     "Failed to get reset GPIO\n");
 
+	ret = of_drm_get_panel_orientation(dev->of_node, &hx->orientation);
+	if (ret < 0) {
+		dev_err(dev, "%pOF: failed to get orientation %d\n", dev->of_node, ret);
+		return ret;
+	}
+
 	/* If the panel is connected on two DSIs then DSI0 left, DSI1 right */
 	dsi_r = of_graph_get_remote_node(dsi->dev.of_node, 1, -1);
 	if (dsi_r) {
