--- a/drivers/phy/rockchip/phy-rockchip-samsung-dcphy.c
+++ b/drivers/phy/rockchip/phy-rockchip-samsung-dcphy.c
@@ -1,50 +1,145 @@
 // SPDX-License-Identifier: GPL-2.0+
 /*
- * Copyright (C) Rockchip Electronics Co.Ltd
+ * Copyright (C) 2025 Rockchip Electronics Co.Ltd
  * Author:
  *      Guochun Huang <hero.huang@rock-chips.com>
  */
 
+#include <dt-bindings/phy/phy.h>
+#include <linux/bitfield.h>
 #include <linux/clk.h>
+#include <linux/hw_bitfield.h>
 #include <linux/init.h>
 #include <linux/kernel.h>
 #include <linux/mfd/syscon.h>
 #include <linux/module.h>
-#include <linux/of_device.h>
+#include <linux/mod_devicetable.h>
+#include <linux/of.h>
 #include <linux/phy/phy.h>
 #include <linux/platform_device.h>
 #include <linux/pm_runtime.h>
 #include <linux/regmap.h>
 #include <linux/reset.h>
 
-#define UPDATE(x, h, l)	(((x) << (l)) & GENMASK((h), (l)))
-#define HIWORD_UPDATE(v, h, l)	(((v) << (l)) | (GENMASK((h), (l)) << 16))
-
 #define BIAS_CON0		0x0000
+#define I_RES_CNTL_MASK		GENMASK(6, 4)
+#define I_RES_CNTL(x)		FIELD_PREP(I_RES_CNTL_MASK, x)
+#define I_RES_059_2UA		I_RES_CNTL(0)
+#define I_RES_100_2UA		I_RES_CNTL(1)
+#define I_RES_094_2UA		I_RES_CNTL(2)
+#define I_RES_113_8UA		I_RES_CNTL(3)
+#define I_RES_089_7UA		I_RES_CNTL(4)
+#define I_RES_111_8UA		I_RES_CNTL(5)
+#define I_RES_108_2UA		I_RES_CNTL(6)
+#define I_RES_120_8UA		I_RES_CNTL(7)
+#define I_DEV_SEL_MASK		GENMASK(1, 0)
+#define I_DEV_SEL(x)		FIELD_PREP(I_DEV_SEL_MASK, x)
+#define I_DEV_DIV_6		I_DEV_SEL(0)
+#define I_DEV_DIV_12		I_DEV_SEL(1)
+#define I_DEV_DIV_20		I_DEV_SEL(2)
+#define I_DEV_DIV_40		I_DEV_SEL(3)
+
 #define BIAS_CON1		0x0004
+#define I_VBG_SEL_MASK		GENMASK(9, 8)
+#define I_VBG_SEL(x)		FIELD_PREP(I_VBG_SEL_MASK, x)
+#define I_VBG_SEL_780MV		I_VBG_SEL(0)
+#define I_VBG_SEL_820MV		I_VBG_SEL(1)
+#define I_VBG_SEL_860MV		I_VBG_SEL(2)
+#define I_VBG_SEL_900MV		I_VBG_SEL(3)
+#define I_BGR_VREF_SEL_MASK	GENMASK(5, 4)
+#define I_BGR_VREF_SEL(x)	FIELD_PREP(I_BGR_VREF_SEL_MASK, x)
+#define I_BGR_VREF_810MV	I_BGR_VREF_SEL(0)
+#define I_BGR_VREF_820MV	I_BGR_VREF_SEL(1)
+#define I_BGR_VREF_830MV	I_BGR_VREF_SEL(2)
+#define I_BGR_VREF_840MV	I_BGR_VREF_SEL(3)
+#define I_LADDER_SEL_MASK	GENMASK(2, 0)
+#define I_LADDER_SEL(x)		FIELD_PREP(I_LADDER_SEL_MASK, x)
+#define I_LADDER_1_00V		I_LADDER_SEL(0)
+#define I_LADDER_0_96V		I_LADDER_SEL(1)
+#define I_LADDER_0_92V		I_LADDER_SEL(2)
+#define I_LADDER_0_88V		I_LADDER_SEL(3)
+#define I_LADDER_0_84V		I_LADDER_SEL(4)
+#define I_LADDER_0_80V		I_LADDER_SEL(5)
+#define I_LADDER_0_76V		I_LADDER_SEL(6)
+#define I_LADDER_0_72V		I_LADDER_SEL(7)
+
+/*
+ * Voltage corrections around reference voltages
+ * The selection between the 400-based or 200-based values for REG_400M
+ * is done by the hw depending on I_MUX below being 400MV or 200MV.
+ */
 #define BIAS_CON2		0x0008
+#define REG_325M_MASK		GENMASK(14, 12)
+#define REG_325M(x)		FIELD_PREP(REG_325M_MASK, x)
+#define REG_325M_295MV		REG_325M(0)
+#define REG_325M_305MV		REG_325M(1)
+#define REG_325M_315MV		REG_325M(2)
+#define REG_325M_325MV		REG_325M(3)
+#define REG_325M_335MV		REG_325M(4)
+#define REG_325M_345MV		REG_325M(5)
+#define REG_325M_355MV		REG_325M(6)
+#define REG_325M_365MV		REG_325M(7)
+#define REG_LP_400M_MASK	GENMASK(10, 8)
+#define REG_LP_400M(x)		FIELD_PREP(REG_LP_400M_MASK, x)
+#define REG_LP_400M_380MV	REG_LP_400M(0)
+#define REG_LP_400M_390MV	REG_LP_400M(1)
+#define REG_LP_400M_400MV	REG_LP_400M(2)
+#define REG_LP_400M_410MV	REG_LP_400M(3)
+#define REG_LP_400M_420MV	REG_LP_400M(4)
+#define REG_LP_400M_430MV	REG_LP_400M(5)
+#define REG_LP_400M_440MV	REG_LP_400M(6)
+#define REG_LP_400M_450MV	REG_LP_400M(7)
+#define REG_400M_MASK		GENMASK(6, 4)
+#define REG_400M(x)		FIELD_PREP(REG_400M_MASK, x)
+#define REG_400M_380MV		REG_400M(0)
+#define REG_400M_390MV		REG_400M(1)
+#define REG_400M_400MV		REG_400M(2)
+#define REG_400M_410MV		REG_400M(3)
+#define REG_400M_420MV		REG_400M(4)
+#define REG_400M_430MV		REG_400M(5)
+#define REG_400M_440MV		REG_400M(6)
+#define REG_400M_450MV		REG_400M(7)
+#define REG_400M_230MV		REG_400M(0)
+#define REG_400M_220MV		REG_400M(1)
+#define REG_400M_210MV		REG_400M(2)
+#define REG_400M_200MV		REG_400M(3)
+#define REG_400M_190MV		REG_400M(4)
+#define REG_400M_180MV		REG_400M(5)
+#define REG_400M_170MV		REG_400M(6)
+#define REG_400M_160MV		REG_400M(7)
+#define REG_645M_MASK		GENMASK(2, 0)
+#define REG_645M(x)		FIELD_PREP(REG_645M_MASK, x)
+#define REG_645M_605MV		REG_645M(0)
+#define REG_645M_625MV		REG_645M(1)
+#define REG_645M_635MV		REG_645M(2)
+#define REG_645M_645MV		REG_645M(3)
+#define REG_645M_655MV		REG_645M(4)
+#define REG_645M_665MV		REG_645M(5)
+#define REG_645M_685MV		REG_645M(6)
+#define REG_645M_725MV		REG_645M(7)
+
 #define BIAS_CON4		0x0010
 #define I_MUX_SEL_MASK		GENMASK(6, 5)
-#define I_MUX_SEL(x)		UPDATE(x, 6, 5)
-#define I_MUX_SEL_400MV		I_MUX_SEL(0)
-#define I_MUX_SEL_200MV		I_MUX_SEL(1)
-#define I_MUX_SEL_530MV		I_MUX_SEL(2)
+#define I_MUX_SEL(x)		FIELD_PREP(I_MUX_SEL_MASK, x)
+#define I_MUX_400MV		I_MUX_SEL(0)
+#define I_MUX_200MV		I_MUX_SEL(1)
+#define I_MUX_530MV		I_MUX_SEL(2)
 
 #define PLL_CON0		0x0100
 #define PLL_EN			BIT(12)
 #define S_MASK			GENMASK(10, 8)
-#define S(x)			UPDATE(x, 10, 8)
+#define S(x)			FIELD_PREP(S_MASK, x)
 #define P_MASK			GENMASK(5, 0)
-#define P(x)			UPDATE(x, 5, 0)
+#define P(x)			FIELD_PREP(P_MASK, x)
 #define PLL_CON1		0x0104
 #define PLL_CON2		0x0108
 #define M_MASK			GENMASK(9, 0)
-#define M(x)			UPDATE(x, 9, 0)
+#define M(x)			FIELD_PREP(M_MASK, x)
 #define PLL_CON3		0x010c
 #define MRR_MASK		GENMASK(13, 8)
-#define MRR(x)			UPDATE(x, 13, 8)
-#define MFR_MASK                GENMASK(7, 0)
-#define MFR(x)			UPDATE(x, 7, 0)
+#define MRR(x)			FIELD_PREP(MRR_MASK, x)
+#define MFR_MASK		GENMASK(7, 0)
+#define MFR(x)			FIELD_PREP(MFR_MASK, x)
 #define PLL_CON4		0x0110
 #define SSCG_EN			BIT(11)
 #define PLL_CON5		0x0114
@@ -52,9 +147,9 @@
 #define PLL_ENABLE_SEL		BIT(8)
 #define PLL_CON6		0x0118
 #define PLL_CON7		0x011c
-#define PLL_LOCK_CNT(x)		UPDATE(x, 15, 0)
+#define PLL_LOCK_CNT(x)		FIELD_PREP(GENMASK(15, 0), x)
 #define PLL_CON8		0x0120
-#define PLL_STB_CNT(x)		UPDATE(x, 15, 0)
+#define PLL_STB_CNT(x)		FIELD_PREP(GENMASK(15, 0), x)
 #define PLL_STAT0		0x0140
 #define PLL_LOCK		BIT(0)
 
@@ -62,46 +157,46 @@
 #define PHY_READY		BIT(1)
 #define PHY_ENABLE		BIT(0)
 #define DPHY_MC_GNR_CON1	0x0304
-#define T_PHY_READY(x)		UPDATE(x, 15, 0)
+#define T_PHY_READY(x)		FIELD_PREP(GENMASK(15, 0), x)
 #define DPHY_MC_ANA_CON0	0x0308
-#define EDGE_CON(x)		UPDATE(x, 14, 12)
-#define EDGE_CON_DIR(x)		UPDATE(x, 9, 9)
+#define EDGE_CON(x)		FIELD_PREP(GENMASK(14, 12), x)
+#define EDGE_CON_DIR(x)		FIELD_PREP(BIT(9), x)
 #define EDGE_CON_EN		BIT(8)
-#define RES_UP(x)		UPDATE(x, 7, 4)
-#define RES_DN(x)		UPDATE(x, 3, 0)
+#define RES_UP(x)		FIELD_PREP(GENMASK(7, 4), x)
+#define RES_DN(x)		FIELD_PREP(GENMASK(3, 0), x)
 #define DPHY_MC_ANA_CON1	0x030c
 #define DPHY_MC_ANA_CON2	0x0310
-#define HS_VREG_AMP_ICON(x)	UPDATE(x, 1, 0)
+#define HS_VREG_AMP_ICON(x)	FIELD_PREP(GENMASK(1, 0), x)
 #define DPHY_MC_TIME_CON0	0x0330
 #define HSTX_CLK_SEL		BIT(12)
-#define T_LPX(x)		UPDATE(x, 11, 4)
+#define T_LPX(x)		FIELD_PREP(GENMASK(11, 4), x)
 #define DPHY_MC_TIME_CON1	0x0334
-#define T_CLK_ZERO(x)		UPDATE(x, 15, 8)
-#define T_CLK_PREPARE(x)	UPDATE(x, 7, 0)
+#define T_CLK_ZERO(x)		FIELD_PREP(GENMASK(15, 8), x)
+#define T_CLK_PREPARE(x)	FIELD_PREP(GENMASK(7, 0), x)
 #define DPHY_MC_TIME_CON2	0x0338
-#define T_HS_EXIT(x)		UPDATE(x, 15, 8)
-#define T_CLK_TRAIL(x)		UPDATE(x, 7, 0)
+#define T_HS_EXIT(x)		FIELD_PREP(GENMASK(15, 8), x)
+#define T_CLK_TRAIL(x)		FIELD_PREP(GENMASK(7, 0), x)
 #define DPHY_MC_TIME_CON3	0x033c
-#define T_CLK_POST(x)		UPDATE(x, 7, 0)
+#define T_CLK_POST(x)		FIELD_PREP(GENMASK(7, 0), x)
 #define DPHY_MC_TIME_CON4	0x0340
-#define T_ULPS_EXIT(x)		UPDATE(x, 9, 0)
+#define T_ULPS_EXIT(x)		FIELD_PREP(GENMASK(9, 0), x)
 #define DPHY_MC_DESKEW_CON0	0x0350
-#define SKEW_CAL_RUN_TIME(x)	UPDATE(x, 15, 12)
+#define SKEW_CAL_RUN_TIME(x)	FIELD_PREP(GENMASK(15, 12), x)
 
-#define SKEW_CAL_INIT_RUN_TIME(x)	UPDATE(x, 11, 8)
-#define SKEW_CAL_INIT_WAIT_TIME(x)	UPDATE(x, 7, 4)
+#define SKEW_CAL_INIT_RUN_TIME(x)	FIELD_PREP(GENMASK(11, 8), x)
+#define SKEW_CAL_INIT_WAIT_TIME(x)	FIELD_PREP(GENMASK(7, 4), x)
 #define SKEW_CAL_EN			BIT(0)
 
 #define COMBO_MD0_GNR_CON0	0x0400
 #define COMBO_MD0_GNR_CON1	0x0404
 #define COMBO_MD0_ANA_CON0	0x0408
-#define COMBO_MD0_ANA_CON1      0x040C
+#define COMBO_MD0_ANA_CON1	0x040c
 #define COMBO_MD0_ANA_CON2	0x0410
 
 #define COMBO_MD0_TIME_CON0	0x0430
 #define COMBO_MD0_TIME_CON1	0x0434
 #define COMBO_MD0_TIME_CON2	0x0438
-#define COMBO_MD0_TIME_CON3	0x043C
+#define COMBO_MD0_TIME_CON3	0x043c
 #define COMBO_MD0_TIME_CON4	0x0440
 #define COMBO_MD0_DATA_CON0	0x0444
 
@@ -113,107 +208,47 @@
 #define COMBO_MD1_TIME_CON0	0x0530
 #define COMBO_MD1_TIME_CON1	0x0534
 #define COMBO_MD1_TIME_CON2	0x0538
-#define COMBO_MD1_TIME_CON3	0x053C
+#define COMBO_MD1_TIME_CON3	0x053c
 #define COMBO_MD1_TIME_CON4	0x0540
 #define COMBO_MD1_DATA_CON0	0x0544
 
 #define COMBO_MD2_GNR_CON0	0x0600
 #define COMBO_MD2_GNR_CON1	0x0604
 #define COMBO_MD2_ANA_CON0	0X0608
-#define COMBO_MD2_ANA_CON1	0X060C
+#define COMBO_MD2_ANA_CON1	0X060c
 #define COMBO_MD2_ANA_CON2	0X0610
 #define COMBO_MD2_TIME_CON0	0x0630
 #define COMBO_MD2_TIME_CON1	0x0634
 #define COMBO_MD2_TIME_CON2	0x0638
-#define COMBO_MD2_TIME_CON3	0x063C
+#define COMBO_MD2_TIME_CON3	0x063c
 #define COMBO_MD2_TIME_CON4	0x0640
 #define COMBO_MD2_DATA_CON0	0x0644
 
 #define DPHY_MD3_GNR_CON0	0x0700
 #define DPHY_MD3_GNR_CON1	0x0704
 #define DPHY_MD3_ANA_CON0	0X0708
-#define DPHY_MD3_ANA_CON1	0X070C
+#define DPHY_MD3_ANA_CON1	0X070c
 #define DPHY_MD3_ANA_CON2	0X0710
 #define DPHY_MD3_TIME_CON0	0x0730
 #define DPHY_MD3_TIME_CON1	0x0734
 #define DPHY_MD3_TIME_CON2	0x0738
-#define DPHY_MD3_TIME_CON3	0x073C
+#define DPHY_MD3_TIME_CON3	0x073c
 #define DPHY_MD3_TIME_CON4	0x0740
 #define DPHY_MD3_DATA_CON0	0x0744
 
-#define T_LP_EXIT_SKEW(x)	UPDATE(x, 3, 2)
-#define T_LP_ENTRY_SKEW(x)	UPDATE(x, 1, 0)
-#define T_HS_ZERO(x)		UPDATE(x, 15, 8)
-#define T_HS_PREPARE(x)		UPDATE(x, 7, 0)
-#define T_HS_EXIT(x)		UPDATE(x, 15, 8)
-#define T_HS_TRAIL(x)		UPDATE(x, 7, 0)
-#define T_TA_GET(x)		UPDATE(x, 7, 4)
-#define T_TA_GO(x)		UPDATE(x, 3, 0)
+#define T_LP_EXIT_SKEW(x)	FIELD_PREP(GENMASK(3, 2), x)
+#define T_LP_ENTRY_SKEW(x)	FIELD_PREP(GENMASK(1, 0), x)
+#define T_HS_ZERO(x)		FIELD_PREP(GENMASK(15, 8), x)
+#define T_HS_PREPARE(x)		FIELD_PREP(GENMASK(7, 0), x)
+#define T_HS_EXIT(x)		FIELD_PREP(GENMASK(15, 8), x)
+#define T_HS_TRAIL(x)		FIELD_PREP(GENMASK(7, 0), x)
+#define T_TA_GET(x)		FIELD_PREP(GENMASK(7, 4), x)
+#define T_TA_GO(x)		FIELD_PREP(GENMASK(3, 0), x)
 
 /* MIPI_CDPHY_GRF registers */
-#define MIPI_DCPHY_GRF_CON0	0x0000
-#define S_CPHY_MODE		HIWORD_UPDATE(1, 3, 3)
-#define M_CPHY_MODE		HIWORD_UPDATE(1, 0, 0)
-
-#define RX_CLK_THS_SETTLE		(0xb30)
-#define RX_LANE0_THS_SETTLE		(0xC30)
-#define RX_LANE0_ERR_SOT_SYNC		(0xC34)
-#define RX_LANE1_THS_SETTLE		(0xD30)
-#define RX_LANE1_ERR_SOT_SYNC		(0xD34)
-#define RX_LANE2_THS_SETTLE		(0xE30)
-#define RX_LANE2_ERR_SOT_SYNC		(0xE34)
-#define RX_LANE3_THS_SETTLE		(0xF30)
-#define RX_LANE3_ERR_SOT_SYNC		(0xF34)
-#define RX_CLK_LANE_ENABLE		(0xB00)
-#define RX_DATA_LANE0_ENABLE		(0xC00)
-#define RX_DATA_LANE1_ENABLE		(0xD00)
-#define RX_DATA_LANE2_ENABLE		(0xE00)
-#define RX_DATA_LANE3_ENABLE		(0xF00)
-
-#define RX_S0C_GNR_CON1			(0xB04)
-#define RX_S0C_ANA_CON1			(0xB0c)
-#define RX_S0C_ANA_CON2			(0xB10)
-#define RX_S0C_ANA_CON3			(0xB14)
-#define RX_COMBO_S0D0_GNR_CON1		(0xC04)
-#define RX_COMBO_S0D0_ANA_CON1		(0xC0c)
-#define RX_COMBO_S0D0_ANA_CON2		(0xC10)
-#define RX_COMBO_S0D0_ANA_CON3		(0xC14)
-#define RX_COMBO_S0D0_ANA_CON6		(0xC20)
-#define RX_COMBO_S0D0_ANA_CON7		(0xC24)
-#define RX_COMBO_S0D0_DESKEW_CON0	(0xC40)
-#define RX_COMBO_S0D0_DESKEW_CON2	(0xC48)
-#define RX_COMBO_S0D0_DESKEW_CON4	(0xC50)
-#define RX_COMBO_S0D0_CRC_CON1		(0xC64)
-#define RX_COMBO_S0D0_CRC_CON2		(0xC68)
-#define RX_COMBO_S0D1_GNR_CON1		(0xD04)
-#define RX_COMBO_S0D1_ANA_CON1		(0xD0c)
-#define RX_COMBO_S0D1_ANA_CON2		(0xD10)
-#define RX_COMBO_S0D1_ANA_CON3		(0xD14)
-#define RX_COMBO_S0D1_ANA_CON6		(0xD20)
-#define RX_COMBO_S0D1_ANA_CON7		(0xD24)
-#define RX_COMBO_S0D1_DESKEW_CON0	(0xD40)
-#define RX_COMBO_S0D1_DESKEW_CON2	(0xD48)
-#define RX_COMBO_S0D1_DESKEW_CON4	(0xD50)
-#define RX_COMBO_S0D1_CRC_CON1		(0xD64)
-#define RX_COMBO_S0D1_CRC_CON2		(0xD68)
-#define RX_COMBO_S0D2_GNR_CON1		(0xE04)
-#define RX_COMBO_S0D2_ANA_CON1		(0xE0c)
-#define RX_COMBO_S0D2_ANA_CON2		(0xE10)
-#define RX_COMBO_S0D2_ANA_CON3		(0xE14)
-#define RX_COMBO_S0D2_ANA_CON6		(0xE20)
-#define RX_COMBO_S0D2_ANA_CON7		(0xE24)
-#define RX_COMBO_S0D2_DESKEW_CON0	(0xE40)
-#define RX_COMBO_S0D2_DESKEW_CON2	(0xE48)
-#define RX_COMBO_S0D2_DESKEW_CON4	(0xE50)
-#define RX_COMBO_S0D2_CRC_CON1		(0xE64)
-#define RX_COMBO_S0D2_CRC_CON2		(0xE68)
-#define RX_S0D3_GNR_CON1		(0xF04)
-#define RX_S0D3_ANA_CON1		(0xF0c)
-#define RX_S0D3_ANA_CON2		(0xF10)
-#define RX_S0D3_ANA_CON3		(0xF14)
-#define RX_S0D3_DESKEW_CON0		(0xF40)
-#define RX_S0D3_DESKEW_CON2		(0xF48)
-#define RX_S0D3_DESKEW_CON4		(0xF50)
+#define MIPI_DCPHY_GRF_CON0		0x0000
+#define S_CPHY_MODE			FIELD_PREP_WM16(BIT(3), 1)
+#define M_CPHY_MODE			FIELD_PREP_WM16(BIT(0), 1)
 
 enum hs_drv_res_ohm {
 	STRENGTH_30_OHM = 0x8,
@@ -257,6 +292,8 @@
 	struct reset_control *apb_rst;
 	struct reset_control *grf_apb_rst;
 	unsigned int lanes;
+	struct phy *phy;
+	u8 type;
 
 	const struct samsung_mipi_dcphy_plat_data *pdata;
 	struct {
@@ -286,6 +323,10 @@
 	u8 hs_settle;
 };
 
+/*
+ * Timing values taken from rk3588 vendor kernel.
+ * Not documented in hw documentation.
+ */
 static const
 struct samsung_mipi_dphy_timing samsung_mipi_dphy_timing_table[] = {
 	{6500, 32, 117, 31, 28, 30, 56, 27, 24, 44, 37},
@@ -935,22 +976,18 @@
 
 static void samsung_mipi_dcphy_bias_block_enable(struct samsung_mipi_dcphy *samsung)
 {
-	u32 bias_con2 = 0x3223;
-
-	regmap_write(samsung->regmap, BIAS_CON0, 0x0010);
-	regmap_write(samsung->regmap, BIAS_CON1, 0x0110);
-	regmap_write(samsung->regmap, BIAS_CON2, bias_con2);
+	regmap_write(samsung->regmap, BIAS_CON0, I_DEV_DIV_6 | I_RES_100_2UA);
+	regmap_write(samsung->regmap, BIAS_CON1, I_VBG_SEL_820MV | I_BGR_VREF_820MV |
+						 I_LADDER_1_00V);
+	regmap_write(samsung->regmap, BIAS_CON2, REG_325M_325MV | REG_LP_400M_400MV |
+						 REG_400M_400MV | REG_645M_645MV);
 
 	/* default output voltage select:
 	 * dphy: 400mv
 	 * cphy: 530mv
 	 */
 	regmap_update_bits(samsung->regmap, BIAS_CON4,
-			   I_MUX_SEL_MASK, I_MUX_SEL_400MV);
-}
-
-static void samsung_mipi_dcphy_bias_block_disable(struct samsung_mipi_dcphy *samsung)
-{
+			   I_MUX_SEL_MASK, I_MUX_400MV);
 }
 
 static void samsung_mipi_dphy_lane_enable(struct samsung_mipi_dcphy *samsung)
@@ -990,11 +1027,27 @@
 
 static void samsung_mipi_dphy_lane_disable(struct samsung_mipi_dcphy *samsung)
 {
+	switch (samsung->lanes) {
+	case 4:
+		regmap_update_bits(samsung->regmap, DPHY_MD3_GNR_CON0,
+				   PHY_ENABLE, 0);
+		fallthrough;
+	case 3:
+		regmap_update_bits(samsung->regmap, COMBO_MD2_GNR_CON0,
+				   PHY_ENABLE, 0);
+		fallthrough;
+	case 2:
+		regmap_update_bits(samsung->regmap, COMBO_MD1_GNR_CON0,
+				   PHY_ENABLE, 0);
+		fallthrough;
+	case 1:
+	default:
+		regmap_update_bits(samsung->regmap, COMBO_MD0_GNR_CON0,
+				   PHY_ENABLE, 0);
+		break;
+	}
+
 	regmap_update_bits(samsung->regmap, DPHY_MC_GNR_CON0, PHY_ENABLE, 0);
-	regmap_update_bits(samsung->regmap, COMBO_MD0_GNR_CON0, PHY_ENABLE, 0);
-	regmap_update_bits(samsung->regmap, COMBO_MD1_GNR_CON0, PHY_ENABLE, 0);
-	regmap_update_bits(samsung->regmap, COMBO_MD2_GNR_CON0, PHY_ENABLE, 0);
-	regmap_update_bits(samsung->regmap, DPHY_MD3_GNR_CON0, PHY_ENABLE, 0);
 }
 
 static void samsung_mipi_dcphy_pll_configure(struct samsung_mipi_dcphy *samsung)
@@ -1059,13 +1112,10 @@
 	timings = samsung_mipi_dphy_timing_table;
 	num_timings = ARRAY_SIZE(samsung_mipi_dphy_timing_table);
 
-	for (i = num_timings; i > 0; i--)
+	for (i = num_timings; i > 1; i--)
 		if (lane_mbps <= timings[i - 1].max_lane_mbps)
 			break;
 
-	if (i == 0)
-		++i;
-
 	return &timings[i - 1];
 }
 
@@ -1294,10 +1344,8 @@
 	samsung_mipi_dphy_clk_lane_timing_init(samsung);
 	samsung_mipi_dphy_data_lane_timing_init(samsung);
 	ret = samsung_mipi_dcphy_pll_enable(samsung);
-	if (ret < 0) {
-		samsung_mipi_dcphy_bias_block_disable(samsung);
+	if (ret < 0)
 		return ret;
-	}
 
 	samsung_mipi_dphy_lane_enable(samsung);
 
@@ -1314,18 +1362,16 @@
 static int samsung_mipi_dcphy_power_on(struct phy *phy)
 {
 	struct samsung_mipi_dcphy *samsung = phy_get_drvdata(phy);
-	enum phy_mode mode = phy_get_mode(phy);
 
-	pm_runtime_get_sync(samsung->dev);
 	reset_control_assert(samsung->apb_rst);
 	udelay(1);
 	reset_control_deassert(samsung->apb_rst);
 
-	switch (mode) {
-	case PHY_MODE_MIPI_DPHY:
+	switch (samsung->type) {
+	case PHY_TYPE_DPHY:
 		return samsung_mipi_dphy_power_on(samsung);
 	default:
-		/* CSI cphy part to be implemented later */
+		/* CPHY part to be implemented later */
 		return -EOPNOTSUPP;
 	}
 
@@ -1335,28 +1381,18 @@
 static int samsung_mipi_dcphy_power_off(struct phy *phy)
 {
 	struct samsung_mipi_dcphy *samsung = phy_get_drvdata(phy);
-	enum phy_mode mode = phy_get_mode(phy);
 
-	switch (mode) {
-	case PHY_MODE_MIPI_DPHY:
+	switch (samsung->type) {
+	case PHY_TYPE_DPHY:
 		samsung_mipi_dphy_lane_disable(samsung);
 		break;
 	default:
-		/* CSI cphy part to be implemented later */
+		/* CPHY part to be implemented later */
 		return -EOPNOTSUPP;
 	}
 
 	samsung_mipi_dcphy_pll_disable(samsung);
-	samsung_mipi_dcphy_bias_block_disable(samsung);
-
-	pm_runtime_put(samsung->dev);
-
-	return 0;
-}
 
-static int samsung_mipi_dcphy_set_mode(struct phy *phy, enum phy_mode mode,
-				       int submode)
-{
 	return 0;
 }
 
@@ -1465,23 +1501,18 @@
 {
 	struct samsung_mipi_dcphy *samsung = phy_get_drvdata(phy);
 
-	pm_runtime_get_sync(samsung->dev);
-
-	return 0;
+	return pm_runtime_resume_and_get(samsung->dev);
 }
 
 static int samsung_mipi_dcphy_exit(struct phy *phy)
 {
 	struct samsung_mipi_dcphy *samsung = phy_get_drvdata(phy);
 
-	pm_runtime_put(samsung->dev);
-
-	return 0;
+	return pm_runtime_put(samsung->dev);
 }
 
 static const struct phy_ops samsung_mipi_dcphy_ops = {
 	.configure = samsung_mipi_dcphy_configure,
-	.set_mode = samsung_mipi_dcphy_set_mode,
 	.power_on  = samsung_mipi_dcphy_power_on,
 	.power_off = samsung_mipi_dcphy_power_off,
 	.init = samsung_mipi_dcphy_init,
@@ -1497,13 +1528,31 @@
 	.max_register = 0x10000,
 };
 
+static struct phy *samsung_mipi_dcphy_xlate(struct device *dev,
+					    const struct of_phandle_args *args)
+{
+	struct samsung_mipi_dcphy *samsung = dev_get_drvdata(dev);
+
+	if (args->args_count != 1) {
+		dev_err(dev, "invalid number of arguments\n");
+		return ERR_PTR(-EINVAL);
+	}
+
+	if (samsung->type != PHY_NONE && samsung->type != args->args[0])
+		dev_warn(dev, "phy type select %d overwriting type %d\n",
+			 args->args[0], samsung->type);
+
+	samsung->type = args->args[0];
+
+	return samsung->phy;
+}
+
 static int samsung_mipi_dcphy_probe(struct platform_device *pdev)
 {
 	struct device *dev = &pdev->dev;
 	struct device_node *np = dev->of_node;
 	struct samsung_mipi_dcphy *samsung;
 	struct phy_provider *phy_provider;
-	struct phy *phy;
 	struct resource *res;
 	void __iomem *regs;
 	int ret;
@@ -1560,17 +1609,17 @@
 		return dev_err_probe(dev, PTR_ERR(samsung->grf_apb_rst),
 				     "Failed to get system grf_apb_rst control\n");
 
-	phy = devm_phy_create(dev, NULL, &samsung_mipi_dcphy_ops);
-	if (IS_ERR(phy))
-		return dev_err_probe(dev, PTR_ERR(phy), "Failed to create MIPI DC-PHY\n");
+	samsung->phy = devm_phy_create(dev, NULL, &samsung_mipi_dcphy_ops);
+	if (IS_ERR(samsung->phy))
+		return dev_err_probe(dev, PTR_ERR(samsung->phy), "Failed to create MIPI DC-PHY\n");
 
-	phy_set_drvdata(phy, samsung);
+	phy_set_drvdata(samsung->phy, samsung);
 
 	ret = devm_pm_runtime_enable(dev);
 	if (ret)
 		return dev_err_probe(dev, ret, "Failed to enable runtime PM\n");
 
-	phy_provider = devm_of_phy_provider_register(dev, of_phy_simple_xlate);
+	phy_provider = devm_of_phy_provider_register(dev, samsung_mipi_dcphy_xlate);
 	if (IS_ERR(phy_provider))
 		return dev_err_probe(dev, PTR_ERR(phy_provider),
 				     "Failed to register phy provider\n");
@@ -1582,8 +1631,8 @@
 {
 	struct samsung_mipi_dcphy *samsung = dev_get_drvdata(dev);
 
-	clk_disable_unprepare(samsung->pclk);
 	clk_disable_unprepare(samsung->ref_clk);
+	clk_disable_unprepare(samsung->pclk);
 
 	return 0;
 }
@@ -1591,9 +1640,20 @@
 static __maybe_unused int samsung_mipi_dcphy_runtime_resume(struct device *dev)
 {
 	struct samsung_mipi_dcphy *samsung = dev_get_drvdata(dev);
+	int ret;
 
-	clk_prepare_enable(samsung->pclk);
-	clk_prepare_enable(samsung->ref_clk);
+	ret = clk_prepare_enable(samsung->pclk);
+	if (ret) {
+		dev_err(samsung->dev, "Failed to enable pclk, %d\n", ret);
+		return ret;
+	}
+
+	ret = clk_prepare_enable(samsung->ref_clk);
+	if (ret) {
+		dev_err(samsung->dev, "Failed to enable reference clock, %d\n", ret);
+		clk_disable_unprepare(samsung->pclk);
+		return ret;
+	}
 
 	return 0;
 }
@@ -1642,13 +1702,13 @@
 static struct platform_driver samsung_mipi_dcphy_driver = {
 	.driver = {
 		.name = "samsung-mipi-dcphy",
-		.of_match_table	= of_match_ptr(samsung_mipi_dcphy_of_match),
+		.of_match_table	= samsung_mipi_dcphy_of_match,
 		.pm = &samsung_mipi_dcphy_pm_ops,
 	},
 	.probe	= samsung_mipi_dcphy_probe,
 };
 module_platform_driver(samsung_mipi_dcphy_driver);
 
-MODULE_AUTHOR("Guochun Huang<hero.huang@rock-chips.com>");
+MODULE_AUTHOR("Guochun Huang <hero.huang@rock-chips.com>");
 MODULE_DESCRIPTION("Samsung MIPI DCPHY Driver");
 MODULE_LICENSE("GPL");
